server/api/muster/muster.controller.ts

```diff
@@ -277,6 +281,135 @@ class MusterController {
+        // Get the unix timestamp of the earliest possible muster window, it could be in the previous week if the
+        // muster window spans the week boundary.
+        const musterTime = moment(muster.startTime, 'HH:mm');
+        let current = moment
+          .unix(since - muster.durationMinutes * 60)
+          .tz(muster.timezone)
+          .startOf('week')
+          .add(musterTime.hour(), 'hours')
+          .add(musterTime.minutes(), 'minutes')
+          .unix();
+        const durationSeconds = muster.durationMinutes * 60;
+        // Loop through each week
+        while (current < until) {
+          // Loop through each day of week to see if any of the windows ended in the query window
+          for (let day = DaysOfTheWeek.Sunday; day <= DaysOfTheWeek.Saturday && current < until; day = nextDay(day)) {
+            const end = current + durationSeconds;
+            // If the window ended in the query window, add it to the list
+            if (end > since && end <= until && dayIsIn(day, muster.days)) {
+              musterWindows.push(buildMusterWindow(unit, current, end, muster));
+            }
+            current += oneDaySeconds;
+          }
+        }
+      }
+    }
+
+    res.json(musterWindows);
+  }
+
+  async getNearestMusterWindow(req: ApiRequest<OrgUnitParams, null, GetNearestMusterWindowQuery>, res: Response) {
+    const timestamp = parseInt(req.query.timestamp);
+
+    const unit = await Unit.findOne({
+      relations: ['org'],
+      where: {
+        id: req.params.unitId,
+        org: req.appOrg!.id,
+      },
+    });
+
+    if (!unit) {
+      throw new NotFoundError('The unit could not be found.');
+    }
+
+    let minDistance: number | null = null;
+    let closestMuster: MusterConfiguration | undefined;
+    let closestStart = 0;
+    let closestEnd = 0;
+
+    for (const muster of unit.musterConfiguration) {
+      if (muster.days === DaysOfTheWeek.None) {
+        continue;
+      }
+      // Get the unix timestamp of the earliest possible muster window in the week of the timestamp
+      const musterTime = moment(muster.startTime, 'HH:mm');
+      let current = moment
+        .unix(timestamp)
+        .tz(muster.timezone)
+        .startOf('week')
+        .add(musterTime.hour(), 'hours')
+        .add(musterTime.minutes(), 'minutes')
+        .unix();
+      const durationSeconds = muster.durationMinutes * 60;
+      // Loop through each day of the week
+      let firstWindowStart: number = 0;
+      let lastWindowStart: number = 0;
+      for (let day = DaysOfTheWeek.Sunday; day <= DaysOfTheWeek.Saturday; day = nextDay(day)) {
+        const end = current + durationSeconds;
+        if (dayIsIn(day, muster.days)) {
+          if (firstWindowStart === 0) {
+            firstWindowStart = current;
+          }
+          lastWindowStart = current;
+          const distanceToWindow = timestamp > end ? timestamp - end : (timestamp < current ? timestamp - current : 0);
```

Comment:
This is pretty tough to read. It also looks very similar to the other `distanceToWindow` bit below. Can we break it into function sort of like this?

```typescript
  function getDistanceToWindow(start: number, end: number, time: number) {
    if (time > end) {
      return time - end;
    }
    if (time < start) {
      return time - start;
    }
    return 0;
  }
```